src/ucp/core/ucp_request.inl

```diff
@@ -318,3 +293,124 @@ static UCS_F_ALWAYS_INLINE void ucp_request_send_tag_stat(ucp_request_t *req)
         UCP_EP_STAT_TAG_OP(req->send.ep, EAGER);
     }
 }
+
+static UCS_F_ALWAYS_INLINE void
+ucp_request_mrail_create(ucp_request_t *req)
+{
+    int i;
+    ucp_rndv_get_mrail_t *mrail;
+
+    ucs_trace_req("mrail create request %p", req);
+    req->send.rndv_get.mrail = (ucp_rndv_get_mrail_t *)ucs_mpool_get_inline(&(req->send.ep->worker)->mrail_mp);
+    req->flags |= UCP_REQUEST_FLAG_RNDV_MRAIL;
+
+    mrail = req->send.rndv_get.mrail;
+
+    for(i = 0; i < ucs_countof(mrail->rail); i++) {
+        req->send.rndv_get.mrail->rail[i].rkey_bundle.rkey = UCT_INVALID_RKEY;
+    }
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_request_mrail_release(ucp_request_t *req)
+{
+    ucs_trace_req("mrail release request %p", req);
+    ucs_mpool_put_inline(req->send.rndv_get.mrail);
+    req->flags &= ~UCP_REQUEST_FLAG_RNDV_MRAIL;
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_request_clear_rails(ucp_dt_state_t *state) {
+    int i;
+    for(i = 0; i < UCP_MAX_RAILS; i++) {
+        state->dt.mrail[i].memh = UCT_MEM_HANDLE_NULL;
+        state->dt.mrail[i].lane = UCP_NULL_LANE;
+    }
+}
+
+static UCS_F_ALWAYS_INLINE int
+ucp_request_is_empty_rail(ucp_dt_state_t *state, int rail) {
+    return state->dt.mrail[rail].lane == UCP_NULL_LANE;
+}
+
+static UCS_F_ALWAYS_INLINE int
+ucp_request_have_rails(ucp_dt_state_t *state) {
+    return !ucp_request_is_empty_rail(state, 0);
+}
+
+static inline int ucp_request_mrail_reg(ucp_request_t *req)
+{
+    ucp_ep_t        *ep      = req->send.ep;
+    ucp_dt_state_t  *state   = &req->send.state.dt;
+    int              cnt     = 0;
+    ucs_status_t     status;
+    int              i;
+    ucp_lane_index_t lane;
+
+    ucs_assert(UCP_DT_IS_CONTIG(req->send.datatype));
+
+    ucp_request_clear_rails(state);
+
+    for (i = 0; i < UCP_MAX_RAILS && ucp_ep_is_rndv_lane_present(ep, i); i++) {
```

Comment:
why not i < ep_config(ep).key<num_rndv_lanes ?

Reply:
updated